### PR TITLE
Improved docker setup

### PR DESCRIPTION
**Pull request by wwelling**
**[VIVO GitHub issue](https://github.com/vivo-project/VIVO/issues)**: https://github.com/vivo-project/VIVO/issues/3939

# What does this pull request do?
This PR provides more configurability for docker compose and build of the VIVO docker image. Primarily addressing port mapping configuration via `.env` file for `docker-compose.yml` variable substitution.

# What's new?

1. solr and tomcat service port mapping configurable via `.env`
2. ability to specify VIVO home path within the container
3. local mount for VIVO solr core data to persist through container restarts

# How should this be tested?

This can be tested by following https://github.com/wwelling/VIVO/blob/issue-3939/README.md#docker-compose. For more exhaustive testing the `.env` file can be modified between container restarts for desired effect.

# Interested parties
@chenejac, @isl3

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1,4 +1,22 @@
-LOCAL_VIVO_HOME=./vivo-home
-RESET_HOME=false
-RESET_CORE=false
-VERBOSE=no
+# Variable substitution for docker-compose.yml
```
Comment (ivanmrsulja) [inline .env:1]: What do you think about separating the .env file into 2 files:
- .env.solr
- .env.tomcat

Would that make it easier to configure for the end user?

File: docker-compose.yml [hunk 2]
```diff
@@ -19,18 +21,26 @@ services:
...
-      - 8080:8080
+      - ${VIVO_HOST_PORT}:${VIVO_CONTAINER_PORT}
     volumes:
-      - ${LOCAL_VIVO_HOME}:/usr/local/vivo/home
+      - ${VIVO_HOST_VIVO_HOME}:${VIVO_CONTAINER_VIVO_HOME}
     networks:
       - vivo
+    depends_on:
+      - solr
+    entrypoint:
```
Comment (ivanmrsulja) [inline docker-compose.yml:38]: Is it a better practice to add a healthcheck as described here (https://medium.com/@nikjos/docker-compose-health-check-for-apache-solr-251c04775e24) to Solr container, and then just say that the depends_on condition for tomcat container is "service_healthy" Solr container? IMO, that would remove the need for custom check script and improve potential integration with orchestration tools if somebody decides to use one.

**Review by ivanmrsulja — COMMENTED** [review]
LGTM and works as intended! I left some comments that are open for discussion, if you have time, so please check them out :slightly_smiling_face:

**Review by isl3 — APPROVED** [review]

**Comment by isl3** [comment]
It might be helpful to some there was a comment somewhere saying VIVO_HOST_PORT/SOLR_HOST_PORT can also specify host IP, but this should be obvious to anyone who knows how docker-compose files work so not an large concern.

File: .env (hunk 1 above)
Comment (wwelling) [reply .env:1]: The `.env` currently is for substituting variables for the `docker-compose.yml` and done without additional docker compose command arguments. These variables are technically not service environment variables unless used to set under `environment` of each service. If necessary, an environment file per service can be added and used for the environment variables of the service. At this time, I opted against env file per service as there is more build time variables (build args, volume, and port variables) needed than actual service environment variables.

File: docker-compose.yml (hunk 2 above)
Comment (wwelling) [reply docker-compose.yml:38]: Yes absolutely. Shortly after I put in this PR I did exactly that with another project with docker compose.

Would this be a reasonable health check for Solr?
```
    healthcheck:
      test: ["CMD-SHELL", "curl -Isf http://localhost:8983/solr/vivocore/admin/ping || exit 1"]
      interval: 15s
      timeout: 5s
      retries: 3
...
    depends_on:
      solr:
        condition: service_healthy
```

Comment (wwelling) [reply docker-compose.yml:38]: FYI, I read that same article but had difficulty with the health check command they used. I will test a bit before adding additional commit.

Comment (ivanmrsulja) [reply docker-compose.yml]: > Yes absolutely. Shortly after I put in this PR I did exactly that for another project with docker compose.
> 
> Would this be a reasonable health check for Solr?
> 
> ```
>     healthcheck:
>       test: ["CMD-SHELL", "curl -Isf http://localhost:8983/solr/vivocore/admin/ping || exit 1"]
>       interval: 15s
>       timeout: 5s
>       retries: 3
> ...
>     depends_on:
>       solr:
>         condition: service_healthy
> ```

This seems like a good approach, nice work.

**Review by ivanmrsulja — APPROVED** [review]
Looks excellent now 🙂 great work!

File: .env (hunk 1 above)
Comment (ivanmrsulja) [reply .env:1]: I understand, then you should leave it as is. I was just wondering what's your opinion and did my suggestion cross your mind at some point during development 😄.

**Review by chenejac — APPROVED** [review]
@wwelling well done. Thanks for this. I don't have any complaint, meaning from my point of view this is ready to be merged.

**Comment by litvinovg** [comment]
@chenejac @ivanmrsulja Can we write down more information about what was tested and reviewed?

**Comment by chenejac** [comment]
> @chenejac @ivanmrsulja Can we write down more information about what was tested and reviewed?

Tested by Ian Slatter, tested and reviewed by Ivan Mrsullja (linux), reviewed by Dragan Ivanovic.

**Pull request merged**